### PR TITLE
config: support v0.3 of the config schema

### PR DESCRIPTION
**Pull request by codeboten**
Adding support for [v0.3.0 release](https://github.com/open-telemetry/opentelemetry-configuration/releases/tag/v0.3.0) of otel configuration. 

As part of the changes, i moved the unmarshaling code into its of file for each format (config_json.go and config_yaml.go) to ensure the resulting struct is consistent. Validated this through unit tests

### CODE REVIEW

**Comment by codeboten** [comment]
More go proxy sadness :(

```
Error: detector.go:14:2: k8s.io/client-go@v0.31.1: Get "https://proxy.golang.org/k8s.io/client-go/@v/v0.31.1.zip": stream error: stream ID 5; PROTOCOL_ERROR
Error: detector.go:15:2: k8s.io/client-go@v0.31.1: Get "https://proxy.golang.org/k8s.io/client-go/@v/v0.31.1.zip": stream error: stream ID 5; PROTOCOL_ERROR
```

File: config/metric.go [hunk 1]
```diff
@@ -208,12 +208,48 @@ func otlpGRPCMetricExporter(ctx context.Context, otlpConfig *OTLPMetric) (sdkmet
...
+	for _, k := range lists.Included {
+		included[attribute.Key(k)] = struct{}{}
+	}
+	excluded := make(map[attribute.Key]struct{})
+	for _, k := range lists.Excluded {
+		if _, ok := included[attribute.Key(k)]; ok {
+			return nil, fmt.Errorf("attribute cannot be in both include and exclude list: %s", k)
+		}
+		excluded[attribute.Key(k)] = struct{}{}
+	}
```
Comment (MrAlias) [inline config/metric.go:237]: nit: you could use the same map for both of these given you can detect the unset state. e.g.

```go
state := make(map[attribute.Key]bool)
// init ^.
allowed, ok := state[k]
if !ok {
	// unknown
	return true
}
return allowed

File: config/metric.go [hunk 2]
```diff
@@ -208,12 +208,48 @@ func otlpGRPCMetricExporter(ctx context.Context, otlpConfig *OTLPMetric) (sdkmet
...
 }
 
+// newIncludeExcludeFilter returns a Filter that includes attributes
+// in the include list and excludes attributes in the excludes list.
+// It returns an error if an attribute is in both lists
+//
+// If IncludeExclude is empty a include-all filter is returned.
+func newIncludeExcludeFilter(lists *IncludeExclude) (attribute.Filter, error) {
+	if lists == nil {
+		return func(kv attribute.KeyValue) bool { return false }, nil
```
[the remark this answers is not in the document]
Comment (codeboten) [reply config/metric.go]: you're right, i'd inverted the test 🤦🏻

[the remark this answers is not in the document]
Comment (codeboten) [reply config/metric.go]: fixed

File: config/metric.go (hunk 1 above)
Comment (codeboten) [reply config/metric.go:237]: Right, but an unset included is different from an unset excluded which is how i ended up with two lists.

**Comment by codeboten** [comment]
The link failure appears unrelated

**Comment by MrAlias** [comment]
> The link check failure appears unrelated

Indeed: https://github.com/open-telemetry/opentelemetry-go-contrib/issues/6183

File: CHANGELOG.md [hunk 3]
```diff
@@ -30,6 +30,10 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
   Use `NewLogProcessor` to configure this setting. (#6116)
 - The deprecated `go.opentelemetry.io/contrib/samplers/aws/xray` package is removed. (#6187)
 
+### Changed
+
+- Updated `go.opentelemetry.io/contrib/config` to use the [v0.3.0](https://github.com/open-telemetry/opentelemetry-configuration/releases/tag/v0.3.0) release of schema which includes backwards incompatible changes. (#5850)
```
Comment (MrAlias) [inline CHANGELOG.md:35]:
```suggestion
- Updated `go.opentelemetry.io/contrib/config` to use the [v0.3.0](https://github.com/open-telemetry/opentelemetry-configuration/releases/tag/v0.3.0) release of schema which includes backwards incompatible changes. (#6126)
```

**Review by MrAlias — APPROVED** [review]

File: CHANGELOG.md (hunk 3 above)
Comment (dmathieu) [reply CHANGELOG.md:36]: This was marked as resolved, but the reference is still not this PR 😸

**Review by dmathieu — APPROVED** [review]

**Pull request merged**